File: server/src/rooms.ts
import { Chess } from 'chess.js';
import type { WebSocket } from 'ws';

export interface Room {
  id: string;
  chess: Chess;
  white: WebSocket | null;
  black: WebSocket | null;
}

// Excludes visually ambiguous characters (0/O, 1/I/L) so a code is easy to
// read aloud or copy correctly.
const ROOM_ID_CHARS = 'ABCDEFGHJKMNPQRSTUVWXYZ23456789';
const ROOM_ID_LENGTH = 6;

const rooms = new Map<string, Room>();

function generateRoomId(): string {
  let id = '';
  for (let i = 0; i < ROOM_ID_LENGTH; i++) {
    id += ROOM_ID_CHARS[Math.floor(Math.random() * ROOM_ID_CHARS.length)];
  }
  return id;
}

export function createRoom(): Room {
  let id = generateRoomId();
  while (rooms.has(id)) id = generateRoomId();
  const room: Room = { id, chess: new Chess(), white: null, black: null };
  rooms.set(id, room);
  return room;
}

export function getRoom(id: string): Room | undefined {
  return rooms.get(id);
}

export function deleteRoom(id: string): void {
  rooms.delete(id);
}

export function roomCount(): number {
  return rooms.size;
}

File: online-server/app/rooms.py
from __future__ import annotations

import random
from dataclasses import dataclass, field

import chess
from fastapi import WebSocket

# Excludes visually ambiguous characters (0/O, 1/I/L) so a code is easy to
# read aloud or copy correctly.
_ROOM_ID_CHARS = "ABCDEFGHJKMNPQRSTUVWXYZ23456789"
_ROOM_ID_LENGTH = 6


@dataclass
class Room:
    id: str
    board: chess.Board = field(default_factory=chess.Board)
    white: WebSocket | None = None
    black: WebSocket | None = None


_rooms: dict[str, Room] = {}


def _generate_room_id() -> str:
    return "".join(random.choice(_ROOM_ID_CHARS) for _ in range(_ROOM_ID_LENGTH))


def create_room() -> Room:
    room_id = _generate_room_id()
    while room_id in _rooms:
        room_id = _generate_room_id()
    room = Room(id=room_id)
    _rooms[room_id] = room
    return room


def get_room(room_id: str) -> Room | None:
    return _rooms.get(room_id)


def delete_room(room_id: str) -> None:
    _rooms.pop(room_id, None)


def room_count() -> int:
    return len(_rooms)
